File: backend/src/domain/repository/lesson-repository.ts
import { LearnedLesson } from "../model/learned-lesson.ts";
import { LearnedLessonDetail } from "../model/learned-lesson-detail.ts";
import { Lesson } from "../model/lesson.ts";

export interface LessonRepository {
  // ================= CATÁLOGO BASE =================
  createLesson(
    course_id: string,
    lesson_order: number,
  ): Promise<{ lesson_id: string }>;

  createLessonTranslation(
    lesson_id: string,
    language_id: string,
    subject: string,
    name: string,
    content_md?: string,
  ): Promise<void>;

  // READ
  readLesson(lesson_id: string, language_id: string): Promise<Lesson | null>;

  // ¡Muy importante para Android! Devuelve todo el curso ordenado
  readLessonsByCourse(
    course_id: string,
    language_id: string,
  ): Promise<Lesson[]>;

  // UPDATE
  updateLessonOrder(lesson_id: string, new_order: number): Promise<void>;

  updateLessonTranslation(
    lesson_id: string,
    language_id: string,
    subject: string,
    name: string,
    content_md?: string,
  ): Promise<void>;

  // DELETE
  deleteLesson(lesson_id: string): Promise<void>;

  // ================= PROGRESO DEL USUARIO ('lessons_learned') =================
  markLessonAsLearned(user_id: string, lesson_id: string): Promise<void>;
  getLearnedLessonsByUser(user_id: string): Promise<LearnedLesson[]>;
  getLearnedLessonsDetailsByUser(
    user_id: string,
    language_id: string,
  ): Promise<LearnedLessonDetail[]>;
  countLessonsByCourse(course_id: string): Promise<number>;
}

File: android/app/src/main/java/com/paralearn/android/domain/repository/LessonRepository.kt
package com.paralearn.android.domain.repository

import com.paralearn.android.domain.model.Lesson

interface LessonRepository {
    suspend fun getCourseLessons(courseId: String): Result<List<Lesson>>
    suspend fun getLesson(lessonId: String): Result<Lesson>
    suspend fun createLesson(courseId: String, lesson: Lesson): Result<Unit>
    suspend fun createLessonTranslation(
        lessonId: String,
        languageName: String,
        name: String,
        subject: String,
        contentMD: String
    ): Result<Unit>
    suspend fun updateLessonOrder(lessonId: String, order: Int): Result<Unit>
    suspend fun updateLessonAlgorithm(lessonId: String, algorithmId: String): Result<Unit>
    suspend fun updateLessonTranslation(
        lessonId: String,
        languageName: String,
        name: String,
        subject: String,
        contentMD: String
    ): Result<Unit>
    suspend fun deleteLesson(lessonId: String): Result<Unit>
}